### PR TITLE
fix: Add schema for merge types

### PR DESCRIPTION
**Pull request by TheRealFalcon**
## Proposed Commit Message

```
fix: Add schema for merge types

Fixes GH-4647
```

## Additional Context
https://github.com/canonical/cloud-init/issues/4647

## Test Steps
Added unit tests

### CODE REVIEW

File: cloudinit/config/schemas/schema-cloud-config-v1.json [hunk 1]
```diff
@@ -421,6 +421,48 @@
...
+        {
+          "type": "string"
+        },
+        {
+          "type": "array",
+          "minItems": 1,
+          "items": {
+            "type": "object",
+            "additionalProperties": false,
+            "properties": {
```
Comment (blackboxsw) [inline cloudinit/config/schemas/schema-cloud-config-v1.json:435]: we need a required definition on these objects
```suggestion
            "required": ["nane", "settings"],
            "properties": {
```

**Review by blackboxsw — CHANGES_REQUESTED** [review]
Let's make the object schema strict by requiring both `name` and `settings` when a dict

**Comment by blackboxsw** [comment]
consolidated comments and a couple tests

```diff
diff --git a/cloudinit/config/schemas/schema-cloud-config-v1.json b/cloudinit/config/schemas/schema-cloud-config-v1.json
index 85d16905b..a553c52ca 100644
--- a/cloudinit/config/schemas/schema-cloud-config-v1.json
+++ b/cloudinit/config/schemas/schema-cloud-config-v1.json
@@ -432,6 +432,10 @@
           "items": {
             "type": "object",
             "additionalProperties": false,
+            "required": [
+              "name",
+              "settings"
+            ],
             "properties": {
               "name": {
                 "type": "string",
diff --git a/tests/unittests/test_merging.py b/tests/unittests/test_merging.py
index c315b5ef6..891031ef0 100644
--- a/tests/unittests/test_merging.py
+++ b/tests/unittests/test_merging.py
@@ -279,6 +279,26 @@ class TestMergingSchema:
                 {"merge_how": [{"name": "list", "settings": "append"}]},
                 "'append' is not of type 'array'",
             ),
+            (
+                {
+                    "merge_how": [
+                        {
+                            "settings": ["recurse_list"],
+                        }
+                    ]
+                },
+                "'name' is a required property",
+            ),
+            (
+                {
+                    "merge_how": [
+                        {
+                            "name": "list",
+                        }
+                    ]
+                },
+                "'settings' is a required property",
+            ),
             (
                 {
                     "merge_how": [
@@ -305,6 +325,17 @@ class TestMergingSchema:
                 },
                 "'badvalue' is not one of",
             ),
+            (
+                {
+                    "merge_how": [
+                        {
+                            "name": "badvalue",
+                            "settings": ["append"],
+                        }
+                    ]
+                },
+                re.escape("'badvalue' is not one of ['list', 'dict', 'str']"),
+            ),
             (
                 {
                     "merge_how": [
```

**Comment by TheRealFalcon** [comment]
Comments applied. Thanks for the patch!

**Review by blackboxsw — APPROVED** [review]
LGTM! Thanks @TheRealFalcon

**Pull request merged**